### PR TITLE
Feature/method reader dont delegate to proxy

### PR DESCRIPTION
**Pull request by JerryShea**
Replaces #454 which was closed in error - see also comments made there

This PR builds upon the following fixes: #456 #457 #458 #472 

This PR stops generated method readers falling back to proxy method readers. If you get a problem with generated method readers then you can always disable them with `-DdisableReaderProxyCodegen`.

It also ensures that problems with generated method readers/writer compilation are logged at warning level and the user is asked to report the problem, and that when we fall back to proxies we see a log warning the user that support will be dropped.

We should now be able to drop proxy method readers/writers in x.25.

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java [hunk 1]
```diff
@@ -243,7 +238,7 @@ public boolean readOne() {
                 context.rollbackOnClose();
         }
 
-        return ok || delegate().readOne();
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java:246]: this should improve performance for when there is nothing to read

File: src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java [hunk 2]
```diff
@@ -316,10 +319,7 @@ private void generateSourceCode() {
                 "throw e;\n" +
                 "}\n" +
                 "catch (Exception e) {\n" +
-                "Jvm.warn().on(this.getClass(), \"Failure to dispatch message, " +
-                "will retry to process without generated code: \" + lastEventName + \"(), " +
-                "bytes: \" + wireIn.bytes().toDebugString(), e);\n" +
-                "return false;\n" +
+                "throw new InvocationTargetRuntimeException(e);\n" +
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java:322]: I'm not 100% sure about throwing this - maybe we should just not catch this exception?

File: src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java (hunk 1 above)
Comment (glukos) [reply src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java:246]: Also this should improve performace if we process "unwanted" event, e.g. if a service processes an event that it's not interested in

**Comment by glukos** [comment]
Looks good

**Review by glukos — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java [hunk 3]
```diff
@@ -260,20 +263,14 @@ private void generateSourceCode() {
         sourceCode.append(eventNameSwitchBlock);
 
         sourceCode.append("default:\n" +
-                "valueIn.skipValue();\n" +
```
Comment (glukos) [inline src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java:263]: I suspect the code above is incorrect now
```
        sourceCode.append("default:\n" +
                "valueIn.skipValue();\n" +
                "return false;\n" +
                "}\n" +
                "}\n" +
                "else {\n" +
                "lastEventName = wireIn.readEvent(String.class);\n" +
                "}\n" +
                "try {\n" +
                "if (Jvm.isDebug())\n" +
                "debugLoggingParselet.accept(lastEventName, valueIn);\n" +
                "if (lastEventName == null)\n" +
                "throw new IllegalStateException(\"Failed to read method name or ID\");\n" +
                "switch (lastEventName) {\n" +
                "case MethodReader.HISTORY:\n" +
                "valueIn.marshallable(messageHistory);\n" +
                "break;\n\n");
```
If we detect `BinaryWireCode.FIELD_NUMBER` and we don't find an appropriate method ID in the switch block, we still return `false` hoping that a delegate will handle this case for us.
Looking at `VanillaMethodReader`, I think we should consume the unexpected event using `WireParser#SKIP_READABLE_BYTES` and return `true`.

File: src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java [hunk 4]
```diff
@@ -243,7 +238,7 @@ public boolean readOne() {
                 context.rollbackOnClose();
```
Comment (glukos) [inline src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java]: I'm not sure that this part with `rollbackOnClose` makes sense when we don't have a delegate.

Normally, the generated instance should never return `false`, but if it happens, we should skip an event (ideally, with a warning) and go on. Rolling back will force the method reader to spin on such "erroneous" event.

**Review by glukos — CHANGES_REQUESTED** [review]
Please see my comments above. This is crucial code for the Services - it would be great to cover "unkown method ID" case with a test.
I think it's also possible to cover "returning `false` by generated code shouldn't halt the progress" with a test using `GeneratingMethodReaderInterceptorReturns` hack - I think it's worth it.

File: src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java (hunk 3 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java:263]: Thanks - let me review again (and add some tests)

File: src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java (hunk 4 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java]: Nice one - well spotted - thanks! Totally agree with you re: comment below and extra tests - this needs to be bulletproof

Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java]: Yep - I have now removed the rollbackOnClose. The method reader always returns true if there is some data to be read as there is always a delegate (there is a default one even if you don't specify - see `MethodReaderDelegationTest#testUnsuccessfulCallNoDelegate()`)

File: src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java (hunk 1 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java:246]: See above comment - we always return true if there was data to be read. Either the service processes it, or the defaultParser does

File: src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java (hunk 3 above)
Comment (JerryShea) [reply src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java:263]: @glukos please re-review!

File: src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java [hunk 5]
```diff
@@ -240,8 +243,9 @@ private void generateSourceCode() {
         sourceCode.append(eventIdSwitchBlock);
 
         sourceCode.append("default:\n" +
-                "valueIn.skipValue();\n" +
-                "return false;\n" +
+                // below should be garbage-free if methodId is low. This will now drop through for defaultParselet
+                "lastEventName = Integer.toString(methodId);\n" +
```
Comment (glukos) [inline src/main/java/net/openhft/chronicle/wire/GenerateMethodReader.java:247]: This is not straightforward, but I guess should work - we'll execute "default" section of method name switch

File: src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java (hunk 1 above)
Comment (glukos) [reply src/main/java/net/openhft/chronicle/wire/AbstractGeneratedMethodReader.java:246]: Yes - this is correct. My point was that services now will process "unexpected" events faster, which is a very real-life scenario. In case e.g. we have a service that's everyone listens to (but each listener is interested in different events).

**Review by glukos — APPROVED** [review]
Looks good, I believe all corner cases are now covered with tests.
